src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -0,0 +1,550 @@
+        try:
+            _upload_assembly_to_workspace(conf, workspace_name, assembly_path, assembly_name)
+        except Exception as e:
+            print(e)
+            failed_names.append(assembly_name)
+
+        counter += 1
+
+    if failed_names:
+        print(f'Failed to upload {failed_names}')
+
+    return failed_names
+
+
+def create_entries_in_ws_and_softlinks_in_csd(
+        csd_upload: str,
+        output_dir: str,
+        assembly_names: list[str],
+        all_assemblies: dict[str, str],
+) -> None:
+    """
+    Create standard entries in workspace and softlinks in collectionssource directory.
+
+    csd_upload: directory of assemblies that have been uploaded to workspace
+    output_dir: output directory to create entries in workspace
+    assembly_names: a list of assembly names that exist in target workspace
+    all_assemblies: a dictionary of assembly name to file path
+    """
+    assembly_ids = create_entries_in_sd_workspace(assembly_names, all_assemblies, output_dir)
+    loader_helper.create_softlinks_in_csd(csd_upload, output_dir, assembly_ids)
+
+
+def main():
+    parser = _get_parser()
+    args = parser.parse_args()
+
+    workspace_id = args.workspace_id
+    kbase_collection = getattr(args, loader_common_names.KBASE_COLLECTION_ARG_NAME)
+    source_version = getattr(args, loader_common_names.SOURCE_VER_ARG_NAME)
+    root_dir = args.root_dir
+    token_filepath = args.token_filepath
+    upload_file_ext = args.upload_file_ext
+    keep_job_dir = args.keep_job_dir
+    overwrite = args.overwrite
+
+    env = args.env
+    kb_base_url = loader_common_names.KB_BASE_URL_MAP[env]
+
+    if workspace_id <= 0:
+        parser.error(f"workspace_id needs to be > 0")
+
+    uid = os.getuid()
+    username = os.getlogin()
+
+    job_dir = loader_helper.make_job_dir(
+        root_dir, loader_common_names.SDK_JOB_DIR, username
+    )
+    csd = loader_helper.make_collection_source_dir(
+        root_dir, loader_common_names.DEFAULT_ENV, kbase_collection, source_version
+    )
+    csd_upload = loader_helper.make_collection_source_dir(
+        root_dir, loader_common_names.DEFAULT_ENV, kbase_collection, source_version, True
+    )
+    output_dir = loader_helper.make_output_dir(
+        root_dir,
+        loader_common_names.SOURCE_DATA_DIR,
+        loader_common_names.WS,
+        env,
+        workspace_id,
+    )
+
+    proc = None
+    conf = None
+
+    try:
+        # start podman service
+        proc = loader_helper.start_podman_service(uid)
+    except Exception as e:
+        raise Exception("Podman service failed to start") from e
+    else:
+        # set up conf, start callback server, and upload assemblies to workspace
+        conf = Conf(job_dir, kb_base_url, token_filepath)
+        workspace_name = conf.ws.get_workspace_info({"id": workspace_id})[1]
+
+        _, uploaded_assembly_names = _read_yaml_file(root_dir, workspace_id, env)
```

Comment:
So there's now one large yaml file in `sourcedata/NCBI` that hold the data for all environments and all uploads, which means worst case there could be > 1M entries in the file. My gut feel is that it'd be better have one file in each genome  subdirectory, even from the standpoint of being able to `cat` a file.
The other issue I see is that it's just a list of what's uploaded, whereas we want to know the target UPA as well

Reply:
Updated:
1. one YAML in each genome subdirectory
2. UPA info is also added

Reply:
:+1: